Add catapult targets for simulated airgapped environment, and kubecf/CAP installation changes to allow this

HartS opened the pull request:
Add catapult targets to "airgap" and "reconnect" a CaaSP cluster.
SSH access to CaaSP nodes is required.

This work assumes `DOCKER_REGISTRY` will only be set when testing
kubecf/CAP installation on an "airgapped" CaaSP cluster (one which is
physically cut off from the internet, or has had internet egress
disabled with the `module-experimental-airgap-up` make target)

Since the airgap simulation leverages the namespaced CiliumNetworkPolicy
resource, the restrictive policies become non-functional when the kubecf
deployment is cleaned. Therefore, `module-experimental-airgap-down` has
been added as a prerequisite of the `kubecf-clean` target. Additionally,
all helm parameters will now be set to ensure no images/resources are
pulled from the internet. Feedback is requested from anyone who is
using catapult with the `DOCKER_REGISTRY` setting overridden

viccuad commented on Makefile:
```diff
@@ -171,7 +179,7 @@ kubecf-build: ##@kubecf Build chart from source and install KubeCF
 	$(MAKE) -C modules/kubecf clean
 
-.PHONY: sc	f
+.PHONY: scf
```
This fix should be:

```suggestion
.PHONY: kubecf
```

viccuad commented on modules/experimental/Makefile:
```diff
@@ -8,3 +8,11 @@ eirini_release:
 	./eirinifs.sh
+
+.PHONY: airgap-up
+airgap-up: airgap-down
```
:100:  calling airgap-down here too

viccuad commented on modules/experimental/airgap_down.sh:
```diff
@@ -0,0 +1,46 @@
+#!/usr/bin/env bash
+
+#!/usr/bin/env bash
```
We have a repeated line here

```suggestion
```

viccuad commented on modules/experimental/airgap_down.sh:
```diff
@@ -0,0 +1,46 @@
+. ../../include/common.sh
+. .envrc
+
+if [[ ${BACKEND} != "caasp4os" ]]; then
```
I would like to use the cap-values `platform` key better, that way it will work if people don't call it with the BACKEND var. See https://github.com/SUSE/catapult/blob/db7da28355dfb2d7034de67bbc50aff36ef396a6/backend/caasp4os/deploy.sh#L120

HartS replied:
`BACKEND` always needs to be set for catapult. That line you linked requires `caasp4os` to be the `BACKEND` in order to generate that configmap. Users don't need to set backend every time they run catapult make targets however, the backend info is in the .envrc  file

HartS replied:
Checking `BACKEND` is more reliable than checking the configmap as well, as you may need to run certain targets when the kubernetes cluster may be unavailable, like kubeconfig, charts, gen-config, cleaning, etc.

viccuad commented on modules/experimental/airgap_up.sh:
```diff
@@ -0,0 +1,46 @@
+. ../../include/common.sh
+. .envrc
+
+if [[ ${BACKEND} != "caasp4os" ]]; then
```
same here, with https://github.com/SUSE/catapult/blob/db7da28355dfb2d7034de67bbc50aff36ef396a6/backend/caasp4os/deploy.sh#L120

viccuad commented on modules/experimental/cilium-block-egress.yaml:
```diff
@@ -0,0 +1,16 @@
+apiVersion: "cilium.io/v2"
```
Maybe this file should go into `catapult/kube/airgap/` to be cleaner, all yaml files are there.

HartS replied:
I wanted to keep experimental work in the experimental folder. Even though I think this is fairly stable now, I think more people would need to use it to be promoted to a stable api feature

viccuad commented on modules/kubecf/install.sh:
```diff
@@ -73,11 +86,16 @@ wait_for "kubectl get crd boshdeployments.quarks.cloudfoundry.org -o name"
 #wait_for "kubectl apply -f ../kube/cf-operator/boshdeployment.yaml --namespace=scf"
 wait_for "kubectl apply -f ../kube/cf-operator/password.yaml --namespace=scf"
-wait_for "kubectl apply -f ../kube/cf-operator/qstatefulset_tolerations.yaml --namespace=scf"
+if [[ "${DOCKER_REGISTRY}" != "registry.suse.com" ]]; then
```
Shouldn't this be as follows? as we don't want to wait if we change the registry, as it will fail.
```suggestion
if [[ "${DOCKER_REGISTRY}" == "registry.suse.com" ]]; then
```

viccuad commented on modules/kubecf/install.sh:
```diff
@@ -73,11 +86,16 @@ wait_for "kubectl get crd boshdeployments.quarks.cloudfoundry.org -o name"
 #wait_for "kubectl delete -f ../kube/cf-operator/boshdeployment.yaml --namespace=scf"
 wait_for "kubectl delete -f ../kube/cf-operator/password.yaml --namespace=scf"
-wait_for "kubectl delete -f ../kube/cf-operator/qstatefulset_tolerations.yaml --namespace=scf"
+if [[ "${DOCKER_REGISTRY}" != "registry.suse.com" ]]; then
```
Same here:
```suggestion
if [[ "${DOCKER_REGISTRY}" == "registry.suse.com" ]]; then
```

viccuad commented on modules/kubecf/install.sh:
```diff
@@ -54,11 +53,25 @@ else
     operator_install_args+=(--set "global.operator.watchNamespace=scf")
 fi
 
+if [[ "${DOCKER_REGISTRY}" != "registry.suse.com" ]]; then
```
Nitpicking: I would move this block after the `echo "Installing CFO"` line.

HartS replied:
I'm unsure about this one. The other args are all set before this is printed, and it should be the last thing that prints before actually the helm chart/release.

viccuad commented on modules/kubecf/install.sh:
```diff
@@ -90,9 +108,20 @@ kubectl create secret generic -n scf susecf-scf.var-cf-admin-password --from-lit
 ## CF-Operator >= 4 don't have deployment name in front of secrets name anymore
 kubectl create secret generic -n scf var-cf-admin-password --from-literal=password="${CLUSTER_PASSWORD}"
 
+kubecf_install_args=(--values scf-config-values.yaml)
```
This concerns me to be honest. So far, we had made a point on saving all the state that we are going to install in the values.yaml, and this departs from it. I think this is painful, and will bring us headaches in the future, both developing and debugging already deployed clusters, because if we get into the habit of setting deployment options inline, you never know on what state the cluster was deployed.

I think these changes should go in `modules/kubecf/gen-config.sh`. Yes, we lose functionality of airgapping "on the fly", but production deployments don't need to be airgapped "on the fly", you configure the deployment to be airgapped, and then you deploy.

HartS commented:
Thanks for the fast turnaround @viccuad . Then I'll check with Prabal to see if he wants to do a quick pass, deal with linting issues, and we can get it merged today

HartS commented:
Addressed PR feedback as well as concerns around how we're handling config customization in general which I discussed with @viccuad directly. The way this is now handled is to apply changes to the config structure by merging "customization" objects as many times as we need to with the jq `*` operator, before finally writing the whole config values back to yaml.
